mitodl/mitxonline
# New design: Course page layout changes

# What are the relevant tickets?

Part of #1773 

# Description (What does it do?)

This does a lot but is focused on the course page redesign:

- Adds feature flagging for all new designs - this will expect a PostHog flag called `jkachel-new-design` if you want to see the changes
- Incorporates the new design for course pages, including:
   - Updates to the global style sheet (site has a white background now)
   - Adds info box for the relevant course details on the right side
   - Displays related program information (with links, though they don't go anywhere right now - program pages aren't in this PR)
   - Adds a tab bar to the course info
   - Adds expandable About This Course section - text after the first paragraph is hidden (and it supports splitting on a tag with a specific class name, which isn't exposed in Wagtail right now)
   - Adds instructor cards with more info modals according to the redesign
   - Adds redesigned upsell message dialog
- Adds some new data to the APIs to allow for these changes
   - Course API now provides for pulling a single course
   - Doing this pulls in some additional data (including program info) that it didn't before
   - Adds API for pulling instructor page details
- Refactors the feature flag code _in the Django app_ to look at Posthog as well as the existing `.env` based feature flags
- Removes the faculty members block from the resource page Wagtail model - this will result in any faculty member data in there being _lost_ so make sure you've converted them to instructor pages first. 

# Screenshots (if appropriate):

Note that this doesn't include mobile layouts - that will be in a separate PR. This is just the desktop view and the feature flag stuff. 

Top half:
![image](https://github.com/mitodl/mitxonline/assets/945611/b0bd28e1-8943-456f-901b-383889b0e879)

Bottom half: 
![image](https://github.com/mitodl/mitxonline/assets/945611/4886fcf6-6a5f-4a6c-83a9-b04df331c09d)

Instructor modal:
![image](https://github.com/mitodl/mitxonline/assets/945611/73992e89-c3ea-4a9a-aa10-64be5d35fd16)

New upsell modal:
![image](https://github.com/mitodl/mitxonline/assets/945611/6b3588c0-7e55-4234-90f9-9b510034843c)

# How can this be tested?

To test, you'll need a Posthog account. (There's a free tier if you're creating one.) Once you have that set up, you will need to put your API token and the host in your `.env` file:

```
POSTHOG_API_HOST=https://app.posthog.com
POSTHOG_API_TOKEN=get your own
```

Then, create a new feature flag called `mitxonline-new-product-page`. The payload can be anything (I set it to True) and make sure to set the Release Condition to roll out to 100% of users:

![image](https://github.com/mitodl/mitxonline/assets/945611/4a9421f9-e40f-435c-8b3d-2e641fcb72dc)

_(I changed the feature flag but have not updated this screenshot so the flag name won't match.)_ 

Finally, before testing, make sure you have some courses set up with a variety of settings - some should be in programs, some should be in more than one program, some shouldn't be in a program, some should have products, some should have financial assistance, etc. Also, _make sure your ad blocker is disabled for your local instance_ - it will likely not be happy about Posthog being there and will then block API calls to it. (This was very much the case for uBlock Origin for me.) 

To test: You can toggle between the old and new designs by *enabling* or *disabling* the flag. (You just need to enable/disable, not adjust the payload.) 

With the flag enabled, you should see the new design. With it disabled, the old design should appear. The new design should match the Figma layout for the course about page (as much as possible - some elements may not appear on a given course depending on what settings they have). The old design should match RC or production. 

Note that switching the flag on or off may not take full effect until you reload a time or two - the React app itself looks for the flag and sometimes browser caching flummoxes that.
